components/lib/nearby/src/main/java/mozilla/components/lib/nearby/NearbyConnection.kt

```diff
@@ -0,0 +1,346 @@
+            private val nearbyConnection: NearbyConnection,
+            /**
+             * The ID of the neighbor, which is not meant for human readability.
+             */
+            val neighborId: String,
+
+            /**
+             * The human readable name of the neighbor.
+             */
+            val neighborName: String,
+
+            /**
+             * A short unique token of printable characters shared by both sides of the
+             * pending connection.
+             */
+            val token: String
+        ) : ConnectionState() {
+            /**
+             * Accepts the connection to the neighbor.
+             */
+            fun accept() {
+                nearbyConnection.connectionsClient.acceptConnection(neighborId, nearbyConnection.payloadCallback)
+                nearbyConnection.updateState(Connecting(neighborId, neighborName))
+            }
+
+            /**
+             * Rejects the connection to the neighbor.
+             */
+            fun reject() {
+                nearbyConnection.connectionsClient.rejectConnection(neighborId)
+                // This should put us back in advertising or discovering.
+                nearbyConnection.updateState(nearbyConnection.connectionState)
+            }
+        }
+
+        /**
+         * The connection has been successfully authenticated (or authentication is disabled).
+         * Unless an error occurs, the next state will be [ReadyToSend].
+         */
+        class Connecting(val neighborId: String, val neighborName: String) : ConnectionState()
+
+        /**
+         * A connection has been made to a neighbor and this device may send a message.
+         * This state is followed by [Sending] or [Failure].
+         */
+        class ReadyToSend(val neighborId: String, val neighborName: String?) : ConnectionState()
+
+        /**
+         * A message is being sent from this device. This state is followed by [ReadyToSend] or
+         * [Failure].
+         */
+        class Sending(val neighborId: String, val neighborName: String?, val payloadId: Long) : ConnectionState()
+
+        /**
+         * A failure has occurred.
+         *
+         * @param message an error message describing the failure
+         */
+        class Failure(val message: String) : ConnectionState()
+    }
+
+    private var connectionsClient: ConnectionsClient = Nearby.getConnectionsClient(context)
+
+    // The is mutated only in updateState(), which can be called from both the main thread and in
+    // callbacks so is synchronized.
+    private lateinit var connectionState: ConnectionState
+
+    // This method is called from both the main thread and callbacks.
+    @Synchronized
+    private fun updateState(cs: ConnectionState) {
+        connectionState = cs
+        listener?.updateState(cs)
+    }
+
+    /**
+     * Starts advertising this device. After calling this, the state will be updated to
```

Comment:
I'd be happy to remove it, but the state changes are externally visible through `NearbyListener.updateState()`.

Reply:
True, I guess that's fair. I left a comment about that further down as well, re using the ObserverRegistry. I feel a bit strongly about following existing patterns since we try to avoid a single listener attached that can leak memory if not removed.

Reply:
Switching to `ObserverRegistry` turned out to be easy-peasy. I haven't set up the lifecycle yet but will do that next.

Reply:
I added a `LifecycleOwner` parameter to `NearbyConnection` which is passed to `register()`. This is provided by `MainActivity` in the sample app. I look forward to hearing whether I do so correctly.